vdr/doc/resolvers.go

```diff
@@ -40,6 +40,60 @@ func (d Resolver) Resolve(id did.DID, metadata *types.ResolveMetadata) (*did.Doc
 	return d.Store.Resolve(id, metadata)
 }
 
+// ResolveControllers accepts a list of documents and finds their controllers
+func (d Resolver) ResolveControllers(input []did.Document) ([]did.Document, error) {
+	// end of the chain
+	if len(input) == 0 {
+		return input, nil
+	}
+
+	var leaves []did.Document
+	var refsToResolve []did.DID
+
+	// for each input document, find its controllers or add the doc itself if its controls itself
+	for _, doc := range input {
+		if len(doc.Controller) == 0 && len(doc.CapabilityInvocation) > 0 {
+			// no controller -> doc is its own controller
+			leaves = append(leaves, doc)
+			continue
+		}
+		for _, ctrlDID := range doc.Controller {
+			if doc.ID.Equals(ctrlDID) {
+				if len(doc.CapabilityInvocation) > 0 {
+					// doc is its own controller
+					leaves = append(leaves, doc)
+				}
+			} else {
+				// add did to be resolved later
+				refsToResolve = append(refsToResolve, ctrlDID)
+			}
+		}
+	}
+	// resolve all unresolved docs
+	for _, ref := range refsToResolve {
+		node, _, err := d.Store.Resolve(ref, nil)
+		if err != nil {
+			return nil, fmt.Errorf("unable to resolve controllers: %w", err)
+		}
+		leaves = append(leaves, *node)
+	}
+
+	// filter deactivated
+	j := 0
+	for i, doc := range leaves {
+		if !isDeactivated(doc) {
+			leaves[i] = doc
```

Comment:
This way of filtering in a memory-optimized way is too smart for its own good and reeks of premature optimization.

`leaves[i] = doc` should be `leaves[j] = doc`

(see committed unit test which now fails)

Reply:
it indeed failed, but is a very common thing within Go